metadata_service/proxy/gremlin_proxy.py

```
@@ -1,29 +1,805 @@
+
+    _label = get_label_from(edge_label)
+
+    edge_traversal = g.V(vertex1_id).outE(_label).where(__.inV().hasId(vertex2_id))
+    for key, value in edge_properties.items():
+        edge_traversal = edge_traversal.has(key, value)
+    edge_id = executor(query=edge_traversal.id(), get=FromResultSet.getOptional)
+
+    # If no link exists, create one!
+    if edge_id is None:
+        g = g.V(vertex1_id).as_('one').V(vertex2_id).addE(_label).from_('one').property('created', timestamp())
+        g = _append_traversal(g, traversal_if_add)
+    else:
+        # Just get the edge in case we need to add properties
+        g = _append_traversal(g.E(edge_id), traversal_if_exists)
+
+    if properties:
+        g = _append_traversal(g, _properties_or_drop_except(edge_label, properties))
+
+    # updated properties, etc.
+    if traversal is not None:
+        g = _append_traversal(g, traversal)
+
+    # return the result
+    return executor(query=g, get=execute if execute is not None else FromResultSet.getOnly)
+
+
+def _expire_other_links(
+        *, executor: ExecuteQuery, g: GraphTraversalSource,
+        edge_label: EdgeTypes, key_property_name: Optional[str] = None,
+        vertex1_label: Optional[Union[str, VertexTypes]] = None, vertex1_key: Optional[str] = None,
+        vertex2_label: Optional[Union[str, VertexTypes]] = None, vertex2_key: Optional[str] = None,
+        vertex1_id: Optional[Union[int, str, P]] = None, vertex2_id: Optional[Union[int, str, P]] = None,
+        edge_direction: Direction = Direction.OUT,
+        **properties: Any) -> None:
+    """
+    Expires edges of a given label other than the ones connecting
+    vertex 1 and 2. If keys are not specified, will expire all edges between the two
+    vertex types
+    """
+
+    assert (vertex1_label is not None) != (vertex1_id is not None), \
```

normally I prefer to raise an exception instead of putting assert in the actual code path.

---

Do you think we could change all the assert to raise exceptions ? That way when it fails, we could see the log as well.